internal/federation/handle.go

```diff
@@ -277,3 +279,105 @@ func HandleMediaRequests(mediaIds map[string]func(w http.ResponseWriter)) func(*
+func HandleTransactionRequests(pduCallback func(gomatrixserverlib.Event), eduCallback func(gomatrixserverlib.EDU)) func(*Server) {
+	return func(srv *Server) {
+		srv.mux.Handle("/_matrix/federation/v1/send/{transactionID}", http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			// Check federation signature
+			fedReq, errResp := gomatrixserverlib.VerifyHTTPRequest(
+				req, time.Now(), gomatrixserverlib.ServerName(srv.ServerName), srv.keyRing,
+			)
+			if fedReq == nil {
+				w.WriteHeader(errResp.Code)
+				b, _ := json.Marshal(errResp.JSON)
+				w.Write(b)
+				return
+			}
+
+			// Unmarshal the request body into a transaction object
+			var transaction gomatrixserverlib.Transaction
+			err := json.Unmarshal(fedReq.Content(), &transaction)
+			if err != nil {
+				log.Printf("Unable to unmarshal transaction body bytes into Transaction object: %s", err.Error())
+				errResp := util.MessageResponse(400, err.Error())
+				w.WriteHeader(errResp.Code)
+				b, _ := json.Marshal(errResp.JSON)
+				w.Write(b)
+				return
+			}
+
+			// Extract the transaction ID from the request vars
+			vars := mux.Vars(req)
+			transaction.TransactionID = gomatrixserverlib.TransactionID(vars["transactionID"])
+
+			// Transactions are limited in size; they can have at most 50 PDUs and 100 EDUs.
+			// https://matrix.org/docs/spec/server_server/latest#transactions
+			if len(transaction.PDUs) > 50 || len(transaction.EDUs) > 100 {
+				log.Printf("Transaction too large. PDUs: %d/50, EDUs: %d/100", len(transaction.PDUs), len(transaction.EDUs))
+				errResp := util.MessageResponse(400, "Transactions are limited to 50 PDUs and 100 EDUs")
+				w.WriteHeader(errResp.Code)
+				b, _ := json.Marshal(errResp.JSON)
+				w.Write(b)
+				return
+			}
+
+			// Construct a response and fill as we process each PDU
+			response := gomatrixserverlib.RespSend{}
+			response.PDUs = make(map[string]gomatrixserverlib.PDUResult, 0)
+			for _, pdu := range transaction.PDUs {
+				var header struct {
+					RoomID string `json:"room_id"`
+				}
+				if err := json.Unmarshal(pdu, &header); err != nil {
+					log.Printf("Transaction '%s': Failed to extract room ID from event", transaction.TransactionID)
+					// We don't know the event ID at this point so we can't return the
+					// failure in the PDU results
+				}
+
+				// Retrieve the room version from the server
+				room := srv.rooms[header.RoomID]
+				if room == nil {
+					continue
+				}
+				roomVersion := gomatrixserverlib.RoomVersion(room.Version)
+
+				event, err := gomatrixserverlib.NewEventFromUntrustedJSON(pdu, roomVersion)
+				if err != nil {
+					// We were unable to verify or process this event.
+					// Tell the sending server that we accepted it anyways, as we're just looking to populate the timeline here
+					log.Printf("Unable to process event '%s' in transaction '%s': %s", event.EventID(), transaction.TransactionID, err.Error())
+				}
```

Comment:
Definitely need to `continue` here and not proceed else we'll panic on `:354`.

Reply:
Derp, good point.